### PR TITLE
Fix moe lora

### PR DESCRIPTION
**Pull request by Jackmin801**
We need to load using meta for moe which makes the fqns and extra lora weights an issue

---

> [!NOTE]
> Extracts a robust `LoRALinear` layer with state_dict hooks and synchronized init, updates DCP loading to strip LoRA/tied weights, and adds comprehensive unit tests.
> 
> - **LoRA layer**:
>   - Move `LoRALinear` to `trainer/models/layers/lora.py` with state_dict pre/post hooks, `__getattr__/__getitem__`, dropout, and generator-seeded init; freezes base params.
> - **Model loading (DCP)**:
>   - `load_dcp_from_hf(model, config, parallel_dims)`: strip `lora_*` from `state_dict`; drop `lm_head.weight` when embeddings are tied; re-init LoRA with DP-replicate-synchronized seed; fix post-meta buffers.
>   - Update caller to pass `parallel_dims`.
> - **LoRA application**:
>   - Import new `LoRALinear`; add debug logging and skip non-linear targets.
>   - Add `strip_lora_from_state_dict` utility.
> - **Tests**:
>   - Add `tests/unit/train/models/test_lora.py` covering state_dict compatibility, loading into/from `nn.Linear`, Sequential/ModuleDict cases, attribute forwarding, and forward behavior.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 02e3de9c389c140b82d09b82ed2f5a9119a5607c. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/prime_rl/trainer/models/layers/lora.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        lora_x = self.lora_dropout(x)
+        lora_output = (lora_x @ self.lora_A.T) @ self.lora_B.T * self.scaling
+        return base_output + lora_output
+
+    def __getattr__(self, name: str) -> Any:
+        """Forward missing attributes to wrapped module."""
+        try:
+            return super().__getattr__(name)  # defer to nn.Module's logic
+        except AttributeError:
+            return getattr(self.base_layer, name)
```
Comment (samsja) [inline src/prime_rl/trainer/models/layers/lora.py:72]: this seems hackky

**Review by samsja — COMMENTED** [review]
I would add a integration test for this, looks a bit dangerous

File: src/prime_rl/trainer/models/layers/lora.py (hunk 1 above)
Comment (Jackmin801) [reply src/prime_rl/trainer/models/layers/lora.py:72]: activation checkpointing in torch does the same thing. I stole it from there

File: src/prime_rl/trainer/models/layers/lora.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        prefix: str,
+        *args: Any,
+    ) -> None:
+        """
+        ``_pre_load_state_dict_hook` is called before ``self._load_from_state_dict()`` is called.
+        For ``LoRALinear``, it will add back the module
+        prefix so that non-LoRALinear modules can be loaded into
+        LoRALinear modules properly.
+        """
+        _replace_by_prefix(state_dict, prefix, prefix + f"{_LORA_PREFIX}")
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/models/layers/lora.py]: yep

File: examples/reverse_text/rl_lora.toml
Comment (mikasenghaas) [inline examples/reverse_text/rl_lora.toml:1]: we have a similar config in `configs/` iirc. let's use this for more experimental/ internal stuff

Comment (Jackmin801) [reply examples/reverse_text/rl_lora.toml:1]: ah ok

**Review by samsja — APPROVED** [review]
lgtm
